### ISSUE TITLE
"This type cannot unbox to a native integer: P6opaque, Whatever" where no WhateverCode is used

### ISSUE THREAD

**Issue by Xliff** [rakudo/rakudo#4820]
## The Problem
Please see the [the following gist](https://gist.github.com/Xliff/74ffdc15f4c20943185f9449240c5e78) for a full listing of the code. It will be reposted here in short order.

## Expected Behavior
The expected behavior is for the given routines to run and return an integer. 

## Actual Behavior

The error message 

```
This type cannot unbox to a native integer: P6opaque, Whatever
  in block <unit> at t/10-raw-hashtable.t line 15
```

is thrown

## Steps to Reproduce
Run the code in the listing.

## Environment

```
uname -a
Linux InfinityV 5.13.0-35-generic #40-Ubuntu SMP Mon Mar 7 08:03:10 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

```
raku -v
Welcome to Rakudo™ v2022.02-108-gcd02552c9.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2022.02-38-g588e00d60.
```

**Comment by Xliff** [comment]
```
===== t/10-raw-hashtable.t =====
use v6;

use NativeCall;

constant gpointer = Pointer;

constant glib      = 'glib-2.0',v0;
constant guint    := uint32;
constant gboolean := uint32;

class GHashTable  is repr<CPointer> { }

my $h = g_hash_table_new(&g_str_hash, &g_str_equal);

my $r = g_hash_table_insert(
  $h,
  toPointer('Jazzy',  typed => Str),
  toPointer('Cheese', typed => Str)
);

say g_hash_table_contains(
  $h,
  toPointer('Jazzy', typed => Str)
);

sub g_hash_table_insert (
  GHashTable $hash_table,
  gpointer   $key,
  gpointer   $value
)
  returns uint32
  is native(glib)
  is export
{ * }

sub g_hash_table_new (
  &hash_func  (Pointer --> guint),
  &equal_func (Pointer $a, Pointer $b --> gboolean)
)
  returns GHashTable
  is native(glib)
  is export
{ * }

sub g_str_equal (gpointer $v1, gpointer $v2)
  returns uint32
  is native(glib)
  is export
{ * }

sub g_str_hash (gpointer $v)
  returns guint
  is native(glib)
  is export
{ * }

sub g_hash_table_contains (GHashTable $hash_table, gpointer $key)
  returns uint32
  is native(glib)
  is export
{ * }

sub toPointer (
  $value,
  :$signed   = False,
  :$double   = False,
  :$direct   = False,
  :$encoding = 'utf8',
  :$all      = False,
  :$typed    = Str
)
  is export
{
  # Properly handle non-Str Cool data.
  return $value if $value ~~ Pointer;
  return ($typed !=== Nil ?? $typed !! Nil) unless $value.defined;
  my ($ov, $use-arr, \t, $v) = ( checkForType($typed, $value), False );
  if $ov ~~ Int && $direct {
    $v = Pointer.new($ov);
  } else {
    given $ov {
      # For all implementor-based classes
      when .^lookup('p').so { $ov .= p }

      when Rat { $ov .= Num; proceed }
      when Int {
        $use-arr = True;
        when $signed.so { t := $double ?? CArray[int64] !!  CArray[int32]  }
        default         { t := $double ?? CArray[uint64] !! CArray[uint32] }
      }
      when Rat | Num {
        $use-arr = True;
        t := $double ?? CArray[num64] !!  CArray[num32]
      }

      # Str
      default {
        $ov = ~$ov unless $ov ~~ Str;
        t := Str;
        # Better to use CArray[uint8] as it is less volatile than Str;
        my $ca = CArray[uint8].new( $ov.encode($encoding) );
        $ov = nativecast(Pointer, $ca);
      }
    }
    if $use-arr {
      $v    = t.new;
      $v[0] = $ov;
    } else {
      $v = $ov;
    }
    $v = nativecast(Pointer, $v) unless $v ~~ Pointer;
  }
  $all.not ?? $v !! ($v, \t);
}

sub resolveNativeType (\T) is export {
  say "Resolving { T.^name } to its Raku equivalent...";
  do given T {
    when num32 | num64     { Num }

    when int8  | uint8  |
         int16 | uint16 |
         int32 | uint32 |
         int64 | uint64
                           { Int }

    when str               { Str }

    default                {
      do if T.REPR eq <CPointer CStruct>.any {
        T
      } else {
        # cw: I don't know if this is the best way to handle this.
        die "Do not know how to handle a type of { .^name }!";
      }
    }
  }
}

sub checkForType(\T, $v is copy) is export {
  if T !=== Nil {
    unless $v ~~ T {
      say "Attempting to convert a { $v.^name } to { T.^name }...";
      my $resolved-name = resolveNativeType(T).^name;
      $resolved-name ~= "[{ T.of.^name }]" if $resolved-name eq 'CArray';
      say "RN: { $resolved-name }";
      if $v.^lookup($resolved-name) -> $m {
        say "Using coercer at { $v.^name }.$resolved-name...";
        $v = $m($v);
      }
      # Note reversal of usual comparison. This is due to the fact that
      # nativecall types must be compatible with the value, not the
      # other way around. In a;; other cases, T and $v should have
      # the same type value.
      die "Value does not support { $v.^name } variables. Will only accept {
        T.^name }!" unless T ~~ $v.WHAT;
    }
  }
  $v;
}
```

**Comment by Xliff** [comment]
Here's a simplified version that works...


```
use v6;

use NativeCall;

constant gpointer = Pointer;

constant glib     = 'glib-2.0',v0;
constant guint    = uint32;
constant gboolean = uint32;

class GHashTable  is repr<CPointer> { }

my $h = g_hash_table_new(
  sub (Str $a) { g_str_hash($a) },
  sub (CArray[uint8] $a, CArray[uint8] $b) { g_str_equal($a, $b) }
);

my Str $j = 'Jazzy';
my $r = g_hash_table_insert($h, $j, 'Cheese');
say $r.so;

say g_hash_table_contains($h, $j);

sub g_hash_table_insert (
  GHashTable $hash_table,
  Str        $key,
  Str        $value
)
  returns uint32
  is native(glib)
  is export
  { * }

sub g_hash_table_new (
  &hash_func  (Str --> guint),
  &equal_func (CArray[uint8] $a, CArray[uint8] $b --> gboolean)
)
  returns GHashTable
  is native(glib)
  is export
  { * }

sub g_str_equal (CArray[uint8] $v1, CArray[uint8] $v2)
  returns uint32
  is native(glib)
  is export
  { * }

sub g_str_hash (Str $v)
  returns guint
  is native(glib)
  is export
  { * }

sub g_hash_table_contains (GHashTable $hash_table, Str $key)
  returns uint32
  is native(glib)
  is export
  { * }

sub malloc (uint64)
  returns Pointer
  is native
  { * }
```

**Comment by MasterDuke17** [comment]
gdb output
```
[dan@alexandria perl6]$ MVM_SPESH_DISABLE=1 gdb --args raku gpointer.raku 
GNU gdb (GDB) 11.2
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from raku...
(gdb) b MVM_exception_throw_adhoc
Function "MVM_exception_throw_adhoc" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (MVM_exception_throw_adhoc) pending.
(gdb) r
Starting program: /home/dan/Source/perl6/install/bin/raku gpointer.raku
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".

Breakpoint 1, MVM_exception_throw_adhoc (tc=0x20000030080, messageFormat=0x7ffff7964990 "This type cannot unbox to a native integer: P6opaque, %s") at src/core/exceptions.c:882
882     MVM_NO_RETURN void MVM_exception_throw_adhoc(MVMThreadContext *tc, const char *messageFormat, ...) {
(gdb) bt
#0  MVM_exception_throw_adhoc (tc=0x20000030080, messageFormat=0x7ffff7964990 "This type cannot unbox to a native integer: P6opaque, %s") at src/core/exceptions.c:882
#1  0x00007ffff77cfd13 in get_int (tc=0x20000030080, st=0x20000960ec0, root=0x20004a9ae60, data=0x20004a9ae78) at src/6model/reprs/P6opaque.c:547
#2  0x00007ffff77bed9f in MVM_repr_get_int (tc=0x20000030080, obj=0x20004a9ae60) at src/6model/reprconv.c:683
#3  0x00007ffff777317d in MVM_nativecall_unmarshal_uint (tc=0x20000030080, value=0x20004a9ae60) at src/core/nativecall.c:266
#4  0x00007ffff7776c57 in callback_handler (cb=0x7ffff7ffa000, cb_args=0x7fffffffbf20, cb_result=0x7fffffffbf18, data=0x20005c76d90) at src/core/nativecall_dyncall.c:404
#5  0x00007ffff7922538 in dcCallback_x64_sysv () from //home/dan/Source/perl6/install/lib/libmoar.so
#6  0x00007ffff492fa1e in g_hash_table_lookup_node (hash_return=<synthetic pointer>, key=0x55555555b550, hash_table=0x55555555bcc0) at ../glib/glib/ghash.c:472
#7  g_hash_table_insert_internal (keep_new_key=0, value=0x55555555b570, key=0x55555555b550, hash_table=0x55555555bcc0) at ../glib/glib/ghash.c:1598
#8  g_hash_table_insert (hash_table=0x55555555bcc0, key=0x55555555b550, value=0x55555555b570) at ../glib/glib/ghash.c:1629
#9  0x00007ffff79215c5 in dcCall_x64_sysv () from //home/dan/Source/perl6/install/lib/libmoar.so
#10 0x00007ffff4e2cb73 in ?? ()
#11 0x00007fffffffc040 in ?? ()
#12 0x00007ffff79213ff in dc_callvm_call_x64 () from //home/dan/Source/perl6/install/lib/libmoar.so
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb) call MVM_dump_backtrace(tc)
   at gpointer.raku:15  (<ephemeral file>:<unit>)
 from gpointer.raku:1  (<ephemeral file>:<unit-outer>)
 from gen/moar/stage2/NQPHLL.nqp:1949  (/home/dan/Source/perl6/install/share/nqp/lib/NQPHLL.moarvm:eval)
 from gen/moar/stage2/NQPHLL.nqp:2154  (/home/dan/Source/perl6/install/share/nqp/lib/NQPHLL.moarvm:evalfiles)
 from gen/moar/stage2/NQPHLL.nqp:2084  (/home/dan/Source/perl6/install/share/nqp/lib/NQPHLL.moarvm:command_eval)
 from gen/moar/Compiler.nqp:111  (/home/dan/Source/perl6/install/share/perl6/lib/Perl6/Compiler.moarvm:command_eval)
 from gen/moar/stage2/NQPHLL.nqp:2039  (/home/dan/Source/perl6/install/share/nqp/lib/NQPHLL.moarvm:command_line)
 from gen/moar/rakudo.nqp:140  (/home/dan/Source/perl6/install/share/perl6/runtime/perl6.moarvm:MAIN)
 from gen/moar/rakudo.nqp:1  (/home/dan/Source/perl6/install/share/perl6/runtime/perl6.moarvm:<mainline>)
 from <unknown>:1  (/home/dan/Source/perl6/install/share/perl6/runtime/perl6.moarvm:<main>)
 from <unknown>:1  (/home/dan/Source/perl6/install/share/perl6/runtime/perl6.moarvm:<entry>)
(gdb) call MVM_dump_bytecode(tc)
00000      getcode            loc_15_obj, Frame_31
00001      capturelex         loc_15_obj
00002      getcode            loc_15_obj, Frame_32
00003      capturelex         loc_15_obj
00004      getcode            loc_15_obj, Frame_34
00005      capturelex         loc_15_obj
00006      getcode            loc_15_obj, Frame_36
00007      capturelex         loc_15_obj
00008      getcode            loc_15_obj, Frame_38
00009      capturelex         loc_15_obj
00010      getcode            loc_15_obj, Frame_40
00011      capturelex         loc_15_obj
00012      getcode            loc_15_obj, Frame_42
00013      capturelex         loc_15_obj
00014      getcode            loc_15_obj, Frame_64
00015      capturelex         loc_15_obj
00016      getcode            loc_15_obj, Frame_81
00017      capturelex         loc_15_obj
00018      getcode            loc_5_obj, Frame_30
00019      takeclosure        loc_5_obj, loc_5_obj
00020      getcode            loc_20_obj, Frame_92
00021      takeclosure        loc_20_obj, loc_20_obj
00022      checkarity         0, 0
00023      paramnamesused     
     annotation: gpointer.raku:1
00024      getlex             loc_2_obj, lex_Frame_29_lex23_obj
00025      decont             loc_3_obj, loc_2_obj
00026      const_s            loc_4_str, 'clone'
00027      dispatch_o         loc_2_obj, 'lang-meth-call', Callsite_1, loc_3_obj, loc_4_str, loc_2_obj
00028      dispatch_v         'raku-capture-lex', Callsite_3, loc_2_obj
00029      bindlex            lex_Frame_29_lex23_obj, loc_2_obj
00030      getcode            loc_3_obj, Frame_31
00031      getcode            loc_6_obj, Frame_32
00032      getcode            loc_7_obj, Frame_34
00033      getcode            loc_8_obj, Frame_36
00034      getcode            loc_9_obj, Frame_38
00035      getcode            loc_10_obj, Frame_40
00036      getcode            loc_11_obj, Frame_42
00037      getcode            loc_12_obj, Frame_64
00038      getcode            loc_13_obj, Frame_81
00039      wval               loc_14_obj, 3, 28
00040      create             loc_14_obj, loc_14_obj
00041      wval               loc_15_obj, 3, 28
00042      wval               loc_16_obj, 2, 47
00043      bindattr_o         loc_14_obj, loc_15_obj, '$!descriptor', loc_16_obj, -1
00044      wval               loc_16_obj, 3, 28
00045      wval               loc_15_obj, 3, 24
00046      bindattr_o         loc_14_obj, loc_16_obj, '$!value', loc_15_obj, -1
00047      set                loc_0_obj, loc_14_obj
00048      wval               loc_14_obj, 3, 28
00049      create             loc_14_obj, loc_14_obj
00050      wval               loc_15_obj, 3, 28
00051      wval               loc_16_obj, 2, 45
00052      bindattr_o         loc_14_obj, loc_15_obj, '$!descriptor', loc_16_obj, -1
00053      wval               loc_16_obj, 3, 28
00054      wval               loc_15_obj, 3, 24
00055      bindattr_o         loc_14_obj, loc_16_obj, '$!value', loc_15_obj, -1
00056      set                loc_1_obj, loc_14_obj
00057      const_s            loc_4_str, '$*CTXSAVE'
00058      getdynlex          loc_14_obj, loc_4_str
00059      set                loc_17_obj, loc_14_obj
00060      isnull             loc_18_int, loc_17_obj
00061      if_i               loc_18_int, label_1(00072)
00062      decont             loc_14_obj, loc_17_obj
00063      const_s            loc_4_str, 'ctxsave'
00064      const_i64_16       loc_19_int, 0
00065      dispatch_o         loc_15_obj, 'lang-find-meth', Callsite_0, loc_14_obj, loc_4_str, loc_19_int
00066      isnull             loc_19_int, loc_15_obj
00067      not_i              loc_19_int, loc_19_int
00068      unless_i           loc_19_int, label_1(00072)
00069      decont             loc_15_obj, loc_17_obj
00070      const_s            loc_4_str, 'ctxsave'
00071      dispatch_v         'lang-meth-call', Callsite_1, loc_15_obj, loc_4_str, loc_17_obj
     annotation: gpointer.raku:11
     label_1:
00072      dispatch_v         'boot-code', Callsite_3, loc_20_obj
     annotation: gpointer.raku:13
00073      set                loc_15_obj, loc_1_obj
00074      getlex             loc_14_obj, lex_Frame_29_lex36_obj
00075      decont             loc_14_obj, loc_14_obj
00076      getlex             loc_21_obj, lex_Frame_29_lex38_obj
00077      set                loc_16_obj, loc_21_obj
00078      ifnonnull          loc_21_obj, label_2(00081)
00079      wval               loc_22_obj, 3, 25
00080      set                loc_16_obj, loc_22_obj
     label_2:
00081      getlex             loc_21_obj, lex_Frame_29_lex37_obj
00082      set                loc_22_obj, loc_21_obj
00083      ifnonnull          loc_21_obj, label_3(00086)
00084      wval               loc_23_obj, 3, 25
00085      set                loc_22_obj, loc_23_obj
     label_3:
00086      dispatch_o         loc_22_obj, 'lang-call', Callsite_22, loc_14_obj, loc_16_obj, loc_22_obj
00087      decont             loc_22_obj, loc_22_obj
00088      dispatch_v         'raku-assign', Callsite_10, loc_15_obj, loc_22_obj
     annotation: gpointer.raku:15
00089      set                loc_15_obj, loc_0_obj
00090      getlex             loc_22_obj, lex_Frame_29_lex35_obj
00091      decont             loc_22_obj, loc_22_obj
00092      const_s            loc_4_str, '&toPointer'
00093      getlexstatic_o     loc_16_obj, loc_4_str
00094      decont             loc_16_obj, loc_16_obj
00095      wval               loc_14_obj, 2, 49
00096      wval               loc_23_obj, 3, 40
00097      dispatch_o         loc_23_obj, 'lang-call', Callsite_23, loc_16_obj, loc_14_obj, loc_23_obj
00098      const_s            loc_4_str, '&toPointer'
00099      getlexstatic_o     loc_14_obj, loc_4_str
00100      decont             loc_14_obj, loc_14_obj
00101      wval               loc_16_obj, 2, 51
00102      wval               loc_21_obj, 3, 40
00103      dispatch_o         loc_21_obj, 'lang-call', Callsite_23, loc_14_obj, loc_16_obj, loc_21_obj
00104      dispatch_o         loc_21_obj, 'lang-call', Callsite_24, loc_22_obj, loc_1_obj, loc_23_obj, loc_21_obj
00105   -> decont             loc_21_obj, loc_21_obj
00106      dispatch_v         'raku-assign', Callsite_10, loc_15_obj, loc_21_obj
     annotation: gpointer.raku:21
00107      const_s            loc_4_str, '&say'
00108      getlexstatic_o     loc_15_obj, loc_4_str
00109      decont             loc_15_obj, loc_15_obj
00110      getlex             loc_21_obj, lex_Frame_29_lex39_obj
00111      decont             loc_21_obj, loc_21_obj
00112      const_s            loc_4_str, '&toPointer'
00113      getlexstatic_o     loc_23_obj, loc_4_str
00114      decont             loc_23_obj, loc_23_obj
00115      wval               loc_22_obj, 2, 49
00116      wval               loc_16_obj, 3, 40
00117      dispatch_o         loc_16_obj, 'lang-call', Callsite_23, loc_23_obj, loc_22_obj, loc_16_obj
00118      dispatch_o         loc_16_obj, 'lang-call', Callsite_22, loc_21_obj, loc_1_obj, loc_16_obj
00119      dispatch_o         loc_16_obj, 'lang-call', Callsite_9, loc_15_obj, loc_16_obj
00120      dispatch_v         'raku-sink', Callsite_3, loc_16_obj
00121      wval               loc_16_obj, 3, 25
00122      return_o           loc_16_obj
(gdb)
```

**Comment by lizmat** [comment]
Wasn't this fixed?

**Comment by artemisia-systems** [comment]
Oh good, I've been running into this same problem for a few hours now and went digging through the issues hoping someone had already documented some executable examples at the very least.

As far as i can tell, this happens anytime you try to pass a native routine as a function pointer to another native routine. For whatever (ha) reason, the dispatching breaks and the stub `{ * }` definition gets executed instead of the native call happening, necessitating the additional layer of indirection that "fixed" the simplified version above. Replacing the sub declarations passed to `g_hash_table_new` directly with the native routines they wrap exhibits the issue again in the simplified example.

I also settled on the same "solution" of wrapping the native calls that needed to be passed as function pointers in a Raku routine which *can* be passed successfully, but there's gotta be a better way to allow you to just grab a native routine as its address when only other native code is going to call into it. None of my attempts to drag that raw pointer back out of a native routine reference have yielded any fruit though.

In my case this mandates inserting raku calls into c<->rust interop that has no need to be slowed down by the orchestrating layer above it. Setting a rust callback in a c library from raku should not necessitate that it is raku specifically that invokes the rust code, and yet that is the position I find myself in.

In summary, I can confirm this is still an extant issue on moar-2023.09 through moar-2024.08 at least, with the known workaround having unfortunate performance implications for realistic usecases.

**Comment by Xliff** [comment]
@artemisia-systems: Nice find. Do you have a working code snippet of your more performant solution lying around somewhere?

**Comment by artemisia-systems** [comment]
```
use NativeCall;

class DLHandle is repr('CPointer') { }
sub dlLoadLibrary(Str --> DLHandle) is native { * }
sub dlFindSymbol(DLHandle, Str --> Pointer) is native { * }

my DLHandle $rs-lib = dlLoadLibrary('./executor-rs/target/release/libexecutor_rs.so');
my $sub-rs-pointer = dlFindSymbol($rs-lib, 'sub');

sub sub-rs(int32 $a, int32 $b) returns int32
    is native('./executor-rs/target/release/executor_rs')
    is symbol('sub') { * }

# no longer necessary in order to pass sub-rs as a function pointer
sub sub-rs-shim(int32 $a, int32 $b) returns int32 {
    sub-rs($a, $b)
}
```

any native function that takes a function pointer and breaks dispatch when passed `&sub-rs` directly can be passed the result from `dlFindSymbol` instead of needing to indirect the rust call through `&sub-rs-shim`, but this is tied to dyncall getting linked in as I haven't been able to get libc `dlopen`/`dlsym` to work the same way.
